packages/mds-ingest-migration-processor/processor.ts

```diff
@@ -30,49 +33,150 @@ import {
 
 const MigrationDataSource: <MigrationSourceEntity>(entity: MigrationEntityType) => StreamSource<MigrationSourceEntity> =
   entity => {
-    const topic = MigrationTopic(SOURCE_TENANT_ID, entity)
+    const topic = MigrationSourceTopic(entity)
     return KafkaSource(topic, { groupId: `${topic}.ingest-migration-processor` })
   }
 
+const MigrationErrorTopic = (entityType: MigrationEntityType) => `${TENANT_ID}.${entityType}.error`
+
 const MigrationErrorSink: <MigrationSourceEntity>(
   entity: MigrationEntityType
 ) => DeadLetterSink<MigrationSourceEntity> = entity => {
-  const topic = `${MigrationTopic(TENANT_ID, entity)}.error`
+  const topic = MigrationErrorTopic(entity)
   return KafkaSink(topic, {
     clientId: `${topic}.ingest-migration-processor`
   })
 }
 
-const DevicesMigrationProcessor = StreamProcessor<Device_v0_4_1 & { id: number }, Device_v1_1_0>(
+const MigratedFrom = (entityType: MigrationEntityType, migrated_from_id: number) => ({
+  migrated_from_source: MigrationSourceTopic(entityType),
+  migrated_from_version: '0.4.1',
+  migrated_from_id
+})
+
+interface MigratedDevice {
+  device: Device_v1_1_0
+  migrated_from: MigratedEntityModel
+}
+
+const IngestServiceMigratedDeviceSink = (): StreamSink<MigratedDevice> => () => {
+  const topic = MigrationSourceTopic('device')
+  return {
+    initialize: async () => undefined,
+    write: async message => {
+      try {
+        const [{ device, migrated_from }] = asArray(message)
+        const migrated = await IngestServiceClient.writeMigratedDevice(device, migrated_from)
+        if (migrated) {
+          logger.info(`Migrated device ${migrated.device_id}`)
+        }
+      } catch (error) {
+        logger.error(`Error migrating device`, { topic, message })
+        throw error
+      }
+    },
+    shutdown: async () => undefined
+  }
+}
+
+const DevicesMigrationProcessor = StreamProcessor<Device_v0_4_1 & IdentityColumn, MigratedDevice>(
   MigrationDataSource('device'),
-  async device => convert_v1_0_0_device_to_1_1_0(convert_v0_4_1_device_to_1_0_0(device)),
-  [],
+  async ({ id, ...device }) => ({
+    device: convert_v1_0_0_device_to_1_1_0(convert_v0_4_1_device_to_1_0_0(device)),
+    migrated_from: MigratedFrom('device', id)
+  }),
+  [IngestServiceMigratedDeviceSink()],
   [MigrationErrorSink('device')]
 )
 
-const EventsMigrationProcessor = StreamProcessor<VehicleEvent_v0_4_1 & { id: number }, VehicleEvent_v1_1_0>(
+interface MigratedVehicleEvent {
+  event: VehicleEvent_v1_1_0
+  migrated_from: MigratedEntityModel
+}
+
+const IngestServiceMigratedVehicleEventSink = (): StreamSink<MigratedVehicleEvent> => () => {
+  const topic = MigrationSourceTopic('event')
+  return {
+    initialize: async () => undefined,
+    write: async message => {
+      try {
+        const [{ event, migrated_from }] = asArray(message)
+        const migrated = await IngestServiceClient.writeMigratedVehicleEvent(event, migrated_from)
+        if (migrated) {
```

Comment:
is there a reason that a `migrated` object might be falsey, but the service call wouldn't throw an error?

Reply:
Yes, if it had been previously migrated offline by some ETL process then the ON CONFLICT DO NOTHING will result in a null value. That’s not an error.